### DISCUSSION TITLE
Changes flux sign after average concentration reaches certain value

### DISCUSSION THREAD

**Discussion by avtarsinghh1991** [idaholab/moose#20731] [Q&A Modules: General] [answered]
Hello MOOSE Experts

I am solving diffusion equation. I want to change the sign of the flux from positive to negative or negative to positive if the average concentration value in the postprocessor calculation reaches certain values. 

Algorithm should be something like this

```
lets say, initially the flux is -ve

for n number_of_cycles
if average_concentration < a 
       flux = +ve
elseif average_concentration > b
       flux  = -ve
end
end
```

Please guide me how should I proceed to put the cycling effect in the code.

Thanks in advance

Best
Avtar

**Comment by GiudGiud** [comment, marked as answer]
Hello

Is this the flux in the volume or at a boundary?
Either way, you will want to work with a ParsedMaterial:
https://mooseframework.inl.gov/source/materials/ParsedMaterial.html

The parsed expression will allow both if statements and using postprocessor values. 

If you find that performance is not satisfactory then you can code that in a regular material as well, you can access postprocessor values from materials, and have the if statement in the material's computeQpProperties routine.

Guillaume

**Reply by avtarsinghh1991** [reply]
Hello Guillaume

Flux is applied at the boundary.

Can you please suggest any example in which if else condition is applied in Parsed Function?

"_If you find that performance is not satisfactory then you can code that in a regular material as well, you can access postprocessor values from materials, and have the if statement in the material's computeQpProperties routine_."

Is there any example for this one as well?

Best
Avtar

**Reply by avtarsinghh1991** [reply]
Edit: I am applying flux as the boundary condition.

 ```
[./flux_bc]
    type = FunctionNeumannBC
    variable = diffused
    boundary = 'flux1'
    function = FluxVal
  [../]
```

Can I still use ParsedMaterial for this one?

**Reply by GiudGiud** [reply]
FunctionNeumannBC works with functions, not material properties.
So you cant use parsed material but you can use ParsedFunction

https://mooseframework.inl.gov/source/functions/MooseParsedFunction.html

**Reply by GiudGiud** [reply]
here the function will be the entire flux, not just the diffusivity btw

**Reply by avtarsinghh1991** [reply]
Thanks Guillaume

Can you please explain the meaning of following if condition.

`Examples include 4e4+1e2*t, sqrt(x*x+y*y+z*z), and if(t<=1.0, 0.1*t, (1.0+0.1)*cos(pi/2*(t-1.0)) - 1.0). `

I understand the first if (t<1.0, ?, ?) but what is the meaning of second and third one.

**Reply by avtarsinghh1991** [reply]
Hello @GiudGiud 

 Here **avg_conc_particle** is the postprocessor function, **Switch** is the factor which I am multiplying with the flux to change its direction. I am getting errors. It seems something is wrong in the following input. Please help.

```
  [./FluxVal_temp]
    type = ParsedFunction
    value = Switch*(Th_Cap/1000)*C_Rate/Faraday
    vars = 'C_Rate Faraday Th_Cap Density'   # theoretical_capacity (Ah/g) *crate (1/h)/F(As/mol)
    vals = '1 96485.3415 300 4750000'
  [../]

 [./Charge_dischage_switch_1]
    type = ParsedFunction
    value = if(avg_conc_particle<10000, Switch = 1)
    vars = 'avg_conc_particle' 
  [../]

  [./Charge_dischage_switch_2]
    type = ParsedFunction
    value = if(avg_conc_particle>10000, Switch = -1)
    vars = 'avg_conc_particle' 
  [../]
```

Best
Avtar

**Reply by GiudGiud** [reply]
So the syntax is `if(condition, value if condition is true, value if condition is false)`

I think you want this:
```
  [Switch]
    type = ParsedFunction
    value = if(avg_conc_particle>10000,  -1, 1)
    vars = 'avg_conc_particle' 
  [../]
```
Switch is then the name of the function you can use elsewhere in the input file

**Reply by avtarsinghh1991** [reply]
@GiudGiud 

Thank you very much. However, I am getting following error

```
*** ERROR ***
/home/asingh/projects/babbler/Isotropic_2D/Isotropic.i:415: expected '=' after parameter name '-1', got ','
```

415 line is 

`value = if(avg_conc_particle>10000,  -1, 1)`

**Comment by GiudGiud** [comment]
needs to be under quotes i think

**Reply by avtarsinghh1991** [reply]
@GiudGiud 

Put under quotes as

` value = if(avg_conc_particle>=10000, '-1', '1')`

still following error

```
*** ERROR ***
/home/asingh/projects/babbler/Isotropic_2D/Isotropic.i:415: invalid character '"' - did you leave a field value blank after a previous '='?
```

**Reply by GiudGiud** [reply]
`value = 'if(avg_conc_particle>=10000, -1, 1)'`

**Comment by avtarsinghh1991** [comment]
It started working with the following

```
  [Switch]
    type = ParsedFunction
    value = 'if(ave_conc_particle>=10000.0, -1, 1)'
    vars = 'ave_conc_particle' 
    vals = 'ave_conc_particle'
  [../]
```

Sorry the comment deleted by mistake.

Thanks

**Comment by avtarsinghh1991** [comment]
@GiudGiud 

Just curious, is it possible to include following algorithm or I have write the source file for it.

```
Switch = -1  % Initial value of Switch

for time_loop
            if (Switch.eq.1.and.ave_conc_particle.ge.50000)then     
                Switch=-1
            elseif (Switch.eq.-1.and.ave_conc_particle.le.10000)then    
                Switch=1                                  
            endif
end

```

Best
Avtar

**Reply by GiudGiud** [reply]
you can nest ifs
`if(t=0, -1, if (ave_conc_particle > 50000, -1, if(ave_conc_particle < 10000, 1, -1)))`

note that some of the logic simplifies out here.
If switch is 1 and ave_conc_particle > 50000
switch = -1
is the same as just
if (ave_conc_particle > 50000)
switch = -1

since if switch is not 1, it is -1 right?

**Reply by avtarsinghh1991** [reply]
Thanks @GiudGiud 

I have a small doubt here.

Lets say the initial value of ave_conc_particle=50000 and switch = -1. Now, I do not want to change switch value unless it reaches 10000. Means, for ave_conc_particle=50000 to 10000, switch = -1. 

For this, `if(avg_conc_particle>=10000, '-1', '1')` condition works absolutely fine.

As soon as, ave_conc_particle=10,000 or less, I need to change switch=1 but only once. It should not change with ave_conc_particle increases from 10000 to 50000.

But with this condition, `if(avg_conc_particle>=10000, '-1', '1')`, as soon as it increases 10,000 switch sign will change again to -1 which I don't want to. I want to keep the Switch = 1 as avg_conc_particle increases from 10,000 to 50,000.

This way, I want to do cycling.

Basically, for ave_conc_particle decreases from 50,000 to 10,000, switch=-1
and for ave_conc_particle increases from 10,000 to 50,000, switch=1.

I hope you got my point. Please let me know if there is any solution for this in MOOSE.

Best
Avtar

**Reply by GiudGiud** [reply]
Hello

Ok I think I see what you mean. You need to retrieve the previous value of the function to perform this. Like this for example:

```
[Functions]
  [function_a]
    type = ParsedFunction
    value = 'if(t=0, -1, if (ave_conc_particle > 50000, -1, if(ave_conc_particle < 10000, 1, old_value)))'
    vars = 'ave_conc_particle old_value'
    vals = 'ave_conc_particle old_value'
  []
[]

[Postprocessors]
  [ave_conc_particle]
...
  []
  [old_value]
    type = FunctionValuePostprocessor
    function = function_a
    execute_on = 'INITIAL TIMESTEP_END'
  []
```

**Reply by avtarsinghh1991** [reply]
Hi Guillaume

Exactly, thank you very much. I will let you know if there will be any issue.

Best
Avtar
